### DISCUSSION TITLE
多集群下helm包部署不了

### DISCUSSION THREAD

**Discussion by SUMMERLm** [kubevela/kubevela#3019] [Q&A] [answered]
基于kubevela，搭建三集群环境，基于helm包无法在多集群下进行部署：
目前已完成多集群下基于pod部署，在主节点下helm包部署应用，但是无法在多集群下helm包的部署
是该功能还不支持，还是我的环境搭建有问题？

**Comment by Somefive** [comment]
You need to install fluxcd addon and enable it in all child clusters. Could you check if the fluxcd controller exists in the child cluster?

**Reply by SUMMERLm** [reply]
kubevela should install in all cluster? i  install kubevela in master k8s cluster，and add other k8s cluster。

PS:
# vela cluster list
CLUSTER        	TYPE           	ENDPOINT
cluster-prod   	X509Certificate	https://192.168.1.65:6443
cluster-staging	X509Certificate	https://192.168.2.235:6443.   //install kubevela this cluster 
cluster-test   	X509Certificate	https://192.168.2.79:6443

The vela-test-yaml as below:

apiVersion: core.oam.dev/v1beta1
kind: Application
metadata:
name: example-app-muli-staging-prod-helm-vela-1216
namespace: default
spec:
components:
- name: redis-comp-muli-test-4
type: helm
properties:
chart: redis-cluster
version: 6.2.7
url: https://charts.bitnami.com/bitnami
repoType: helm
policies:
- name: example-multi-env-policy
type: env-binding
properties:
envs:
- name: staging
placement:
clusterSelector:
name: cluster-staging
selector:
components:
- redis-comp-muli-test-4
- name: prod
placement:
clusterSelector:
name: cluster-prod
selector:
components:
- redis-comp-muli-test-4

workflow:
steps:
- name: deploy-staging
type: deploy2env
properties:
policy: example-multi-env-policy
env: staging
- name: deploy-prod
type: deploy2env
properties:
policy: example-multi-env-policy
env: prod
}

I can see redis deploy in cluster-staging, but  cluster-prod have nothing. 
and no logs to trace.
# vela logs app-delivering-chart-nginx-3
Error: no resources in the application deployed yet

**Reply by Somefive** [reply]
No, not install kubevela in all clusters. I mean the fluxcd addon should be enabled in the child cluster which you want to deploy helm component to.

**Reply by Somefive** [reply]
Child cluster do not need to install kubevela controller. But depending on the functions you would like to use in child cluster, you need to install corresponding dependencies. Such as fluxcd addon or RolloutController.

**Reply by SUMMERLm** [reply]
master cluster:
[root@test]# vela addon list
NAME                      	DESCRIPTION                                                                  	STATUS
fluxcd                    	Flux is a set of continuous and progressive delivery solutions for Kubernetes	installed
istio                     	istio Controller is a Kubernetes Controller for manage traffic.              	uninstalled
keda                      	KEDA is a Kubernetes-based Event Driven Autoscaler.                          	uninstalled
kruise                    	Kruise is a Kubernetes extended suite for application automations            	uninstalled
observability             	An out of the box solution for KubeVela observability                        	uninstalled
ocm-cluster-manager       	ocm-cluster-manager can deploy an OCM hub cluster environment.               	uninstalled
prometheus                	Prometheus is an open-source systems monitoring and alerting toolkit         	uninstalled
terraform-alibaba         	Kubernetes Terraform Controller for Alibaba Cloud                            	uninstalled
terraform-aws             	Kubernetes Terraform Controller for AWS                                      	uninstalled
terraform-azure           	Kubernetes Terraform Controller for Azure                                    	uninstalled
terraform/provider-alibaba	Kubernetes Terraform Controller for Alibaba Cloud                            	uninstalled
terraform/provider-aws    	Kubernetes Terraform Controller for AWS                                      	uninstalled
terraform/provider-azure  	Kubernetes Terraform Controller for Azure                                    	uninstalled

**Reply by Somefive** [reply]
If you previously enabled the fluxcd addon, you can find an application called addon-fluxcd in vela-system. You can verify the content in it. Check the workflow spec and the appliedResource records in application status.

**Reply by SUMMERLm** [reply]
I need to install fluxcd in all child cluster?

**Reply by Somefive** [reply]
You can install fluxcd in child cluster by using *vela addon enable fluxcd*. But since this feature is introduced recently, you need to ensure the vela CLI tool is the latest revision.

**Reply by SUMMERLm** [reply]
[root@host-192-168-2-235 test]# kubectl get all -n vela-system
NAME                                                READY   STATUS    RESTARTS        AGE
pod/helm-test-express-server-7bc47f4cff-9sp42       1/1     Running   0               23h
pod/kubevela-application-test                       0/1     Error     0               23h
pod/kubevela-cluster-gateway-5b6c7f84db-mn8vp       1/1     Running   0               23h
pod/kubevela-vela-core-5d76bc9bdc-nbmhg             1/1     Running   10 (100m ago)   23h
pod/kubevela-vela-core-apiserver-5b9b859c48-w5mw8   1/1     Running   0               23h
pod/terraform-controller-688f56b44b-25l46           1/1     Running   0               23h

NAME                                       TYPE        CLUSTER-IP       EXTERNAL-IP   PORT(S)    AGE
service/helm-test-express-server           ClusterIP   10.111.149.215   <none>        8000/TCP   23h
service/kubevela-cluster-gateway-service   ClusterIP   10.108.73.120    <none>        9443/TCP   23h
service/kubevela-vela-core-apiserver       ClusterIP   10.101.70.61     <none>        80/TCP     23h
service/vela-core-webhook                  ClusterIP   10.111.115.165   <none>        443/TCP    23h

NAME                                           READY   UP-TO-DATE   AVAILABLE   AGE
deployment.apps/helm-test-express-server       1/1     1            1           23h
deployment.apps/kubevela-cluster-gateway       1/1     1            1           23h
deployment.apps/kubevela-vela-core             1/1     1            1           23h
deployment.apps/kubevela-vela-core-apiserver   1/1     1            1           23h
deployment.apps/terraform-controller           1/1     1            1           23h

NAME                                                      DESIRED   CURRENT   READY   AGE
replicaset.apps/helm-test-express-server-7bc47f4cff       1         1         1       23h
replicaset.apps/kubevela-cluster-gateway-5b6c7f84db       1         1         1       23h
replicaset.apps/kubevela-vela-core-5d76bc9bdc             1         1         1       23h
replicaset.apps/kubevela-vela-core-apiserver-5b9b859c48   1         1         1       23h
replicaset.apps/terraform-controller-688f56b44b           1         1         1       23h

**Reply by SUMMERLm** [reply]
>1.1.13

**Reply by Somefive** [reply]
I notice that you use v1.1 which does not support this feature yet. We recommend you to upgrade to v1.2. If you want to use v1.1, youcan manually install fluxcd in child cluster to make it work as well.

**Reply by SUMMERLm** [reply]
Thks. i'll try

**Reply by SUMMERLm** [reply]
Hi five：i can not enable addon fluxcd

[root@node1 vela_helm_controller]# vela addon enable fluxcd
I1230 10:23:04.653864   44619 apply.go:106] "patching object" name="addon-fluxcd" resource="core.oam.dev/v1beta1, Kind=Application"
I1230 10:23:05.575142   44619 apply.go:106] "patching object" name="helm" resource="core.oam.dev/v1beta1, Kind=ComponentDefinition"
I1230 10:23:05.598811   44619 apply.go:106] "patching object" name="kustomize-json-patch" resource="core.oam.dev/v1beta1, Kind=TraitDefinition"
I1230 10:23:05.608049   44619 apply.go:106] "patching object" name="kustomize-patch" resource="core.oam.dev/v1beta1, Kind=TraitDefinition"
I1230 10:23:05.617159   44619 apply.go:106] "patching object" name="kustomize" resource="core.oam.dev/v1beta1, Kind=ComponentDefinition"
I1230 10:23:05.638364   44619 apply.go:106] "patching object" name="kustomize-strategy-merge" resource="core.oam.dev/v1beta1, Kind=TraitDefinition"
I1230 10:23:05.646441   44619 apply.go:106] "patching object" name="component-uischema-helm" resource="/v1, Kind=ConfigMap"
⠼ Waiting addon application running. It is now in phase: runningWorkflow (timeout 251/600 seconds)...

**Reply by Somefive** [reply]
Try to uninstall fluxcd first and then re-enable it.

**Comment by barnettZQG** [comment]
@SUMMERLm 说明一下 KubeVela 的版本，请使用1.2.0-rc.1版本支持多集群下部署helm。

**Reply by SUMMERLm** [reply]
版本：1.1.13。之前用1.11也试过。
环境问题目前已经前置验证：1:各子集群下都已安装helm，并完成基于helm进行当前单集群下测试包的部署。2：可以完成多集群下的pod部署。
3:通过vela在多集群下基于helm部署时，只会在vela所在的集群上部署上，但是其他集群没有反应。

**Reply by barnettZQG** [reply]
@SUMMERLm 1.1 版本中 helm还不支持多集群部署，请测试使用 1.2.0-rc.1版本。

**Reply by SUMMERLm** [reply]
好的，多谢.

**Reply by SUMMERLm** [reply]
请问1.2.0-rc.1版本也需要addon fluxcd enable吗？目前一直无法enable该组件

[root@node1 vela_helm_controller]# vela addon enable fluxcd
I1230 10:23:04.653864   44619 apply.go:106] "patching object" name="addon-fluxcd" resource="core.oam.dev/v1beta1, Kind=Application"
I1230 10:23:05.575142   44619 apply.go:106] "patching object" name="helm" resource="core.oam.dev/v1beta1, Kind=ComponentDefinition"
I1230 10:23:05.598811   44619 apply.go:106] "patching object" name="kustomize-json-patch" resource="core.oam.dev/v1beta1, Kind=TraitDefinition"
I1230 10:23:05.608049   44619 apply.go:106] "patching object" name="kustomize-patch" resource="core.oam.dev/v1beta1, Kind=TraitDefinition"
I1230 10:23:05.617159   44619 apply.go:106] "patching object" name="kustomize" resource="core.oam.dev/v1beta1, Kind=ComponentDefinition"
I1230 10:23:05.638364   44619 apply.go:106] "patching object" name="kustomize-strategy-merge" resource="core.oam.dev/v1beta1, Kind=TraitDefinition"
I1230 10:23:05.646441   44619 apply.go:106] "patching object" name="component-uischema-helm" resource="/v1, Kind=ConfigMap"
⠼ Waiting addon application running. It is now in phase: runningWorkflow (timeout 251/600 seconds)...

**Reply by SUMMERLm** [reply]
之前enable1.1版本fluxcd是可以的

**Reply by chivalryq** [reply]
是卡在runningWorkflow 状态了吗？用vela addon status fluxcd查看一下状态？

**Reply by SUMMERLm** [reply]
[root@node1 vela-helm-multicluster]# vela addon status fluxcd
addon fluxcd status is enabling
diagnose addon info from application addon-fluxcdAbout:

  Name:      	addon-fluxcd
  Namespace: 	vela-system
  Created at:	2021-12-30 09:39:54 +0800 CST
  Status:    	runningWorkflow

Workflow:

  mode: StepByStep
  finished: false
  Suspend: false
  Terminated: false
  Steps
  - id:guytg78bfy
    name:deploy-control-plane
    type:apply-application
    phase:succeeded
    message:
  - id:bjr4gc7t9x
    name:deploy-runtime
    type:deploy2runtime
    phase:failed
    message:step app: step apply: step cluster-glouble-ux-flux-system-namespace: run step(provider=oam,do=component-apply): DispatchStandardWorkload: cannot apply manifest, name: flux-system apiVersion: v1 kind: Namespace: cannot apply ApplyOption: existing object is managed by other application fluxcd

**Comment by chivalryq** [comment]
Oh I see, maybe you should uninstall fluxcd in sub-clusters and clean up. Then disable fluxcd and enable again. 

Because fluxcd addon in v1.2 will dispatch some resource in sub-clusters which may cause some confiict.

And vela-core needn't to be installed in sub-clusters neither( I understand you maybe wanna enable them respectively in every cluster)

**Reply by chivalryq** [reply]
let me know if this work ^^

**Reply by SUMMERLm** [reply]
1）子集群的fluxcd已disable。
# vela addon list
NAME                      	DESCRIPTION                                                                  	STATUS
fluxcd                    	Flux is a set of continuous and progressive delivery solutions for Kubernetes	uninstalled



2）主集群下disable fluxcd出现问题如下：
#vela addon disable fluxcd。
# vela addon list
NAME               	REGISTRY	DESCRIPTION                                                                                                                         	STATUS
fluxcd             	KubeVela	Extended workload to do continuous and progressive delivery                                                                         	disabling

fluxcd一直处于disabling状态。
  mode: StepByStep
  finished: false
  Suspend: false
  Terminated: false
  Steps
  - id:guytg78bfy
    name:deploy-control-plane
    type:apply-application
    phase:succeeded
    message:
  - id:bjr4gc7t9x
    name:deploy-runtime
    type:deploy2runtime
    phase:failed
    message:step app: step apply: step cluster-glouble-flux-system-namespace: run step(provider=oam,d
o=component-apply): DispatchStandardWorkload: cannot apply manifest, name: flux-system apiVersion: v1
 kind: Namespace: cannot apply ApplyOption: existing object is managed by other application fluxcd

**Reply by Somefive** [reply]
Could you run `kubectl get application addon-fluxcd -n vela-system -oyaml` and show the full details of the application?

**Reply by SUMMERLm** [reply]
I want to confirm that vela v1.2.0 is ok to  install app by helm  in multicluster ?

PS:my env is now crash , I follow the https://kubevela.io/zh/docs/next/install to uninstall vela ,but too many error. like crd cannot delete,  cr cannnot delete , addon cannot disable etc..
i need to build new test env to try。

**Reply by Somefive** [reply]
Yes. Vela 1.2 supports it.

**Reply by Somefive** [reply]
Uninstalling with helm can be risky. The kubevela controller could be uninstalled first, then applications cannot be deleted normally since there is no controller to handle them.

**Reply by SUMMERLm** [reply]
got it , thks.

**Comment by SUMMERLm** [comment]
Hi Five:
    I install new env like below, vela addon fluxcd enabled
   but still can not use helm to intall app in other cluster(my env is glouble cluster):

**### 1: vela clutser**
[root@kubevela-master vela-helm-multicluster]# helm version
version.BuildInfo{Version:"v3.7.2", GitCommit:"663a896f4a815053445eec4153677ddc24a0a361", GitTreeState:"clean", GoVersion:"go1.16.10"}

[root@kubevela-master vela-helm-multicluster]# helm repo list
NAME    	URL
kubevela	https://charts.kubevela.net/core
bitnami 	https://charts.bitnami.com/bitnami

[root@kubevela-master vela-helm-multicluster]# kubectl version
Client Version: version.Info{Major:"1", Minor:"23", GitVersion:"v1.23.1", GitCommit:"86ec240af8cbd1b60bcc4c03c20da9b98005b92e", GitTreeState:"clean", BuildDate:"2021-12-16T11:41:01Z", GoVersion:"go1.17.5", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"23", GitVersion:"v1.23.1", GitCommit:"86ec240af8cbd1b60bcc4c03c20da9b98005b92e", GitTreeState:"clean", BuildDate:"2021-12-16T11:34:54Z", GoVersion:"go1.17.5", Compiler:"gc", Platform:"linux/amd64"}


[root@kubevela-master vela-helm-multicluster]# vela version
Version: v1.2.0-rc.2
GitRevision: git-f468814
GolangVersion: go1.16.12

[root@kubevela-master vela-helm-multicluster]# vela cluster list
CLUSTER        	TYPE           	ENDPOINT
cluster-glouble	X509Certificate	https://192.168.1.153:6443
cluster-local  	X509Certificate	https://192.168.2.165:6443
glouble-ux     	X509Certificate	https://192.168.1.153:6443

[root@kubevela-master vela-helm-multicluster]# vela addon list
NAME               	REGISTRY	DESCRIPTION                                                                                                                         	STATUS
fluxcd             	KubeVela	Extended workload to do continuous and progressive delivery                                                                         	**enabled**
velaux             	KubeVela	The KubeVela User Experience (UX ). Dashboard Designed as an extensible, application-oriented delivery and management control panel.	**enabled**
terraform-alibaba  	KubeVela	Kubernetes Terraform Controller for Alibaba Cloud                                                                                   	disabled
terraform-aws      	KubeVela	Kubernetes Terraform Controller for AWS                                                                                             	disabled
terraform-azure    	KubeVela	Kubernetes Terraform Controller for Azure                                                                                           	disabled
ocm-cluster-manager	KubeVela	ocm-cluster-manager can deploy an OCM hub cluster environment.                                                                      	disabled
terraform          	KubeVela	Terraform Controller is a Kubernetes Controller for Terraform.                                                                      	disabled

[root@kubevela-master vela-helm-multicluster]# cat vela-helm-multicluster-test-glouble.yaml
apiVersion: core.oam.dev/v1beta1
kind: Application
metadata:
  name: example-app-helm-muli-glouble
  namespace: default
spec:
  components:
    - name: redis-comp-muli-glouble
      type: helm
      properties:
        chart: redis-cluster
        version: 6.2.7
        url: https://charts.bitnami.com/bitnami
        repoType: helm
  policies:
    - name: example-multi-env-policy
      type: env-binding
      properties:
        envs:
          - name: glouble
            placement:
              clusterSelector:
                name: cluster-glouble
            selector:
              components:
                - redis-comp-muli-glouble

  workflow:
    steps:
      - name: deploy-glouble
        type: deploy2env
        properties:
          policy: example-multi-env-policy
          env: glouble

[root@kubevela-master vela-helm-multicluster]# vela logs example-app-helm-muli-glouble
Error: no resources in the application deployed yet


### **2: glouble cluster**
[root@vela-master ~]# helm version
version.BuildInfo{Version:"v3.7.2", GitCommit:"663a896f4a815053445eec4153677ddc24a0a361", GitTreeState:"clean", GoVersion:"go1.16.10"}
 
[root@vela-master ~]# helm repo list
NAME   	URL                               
bitnami	https://charts.bitnami.com/bitnami

[root@vela-master ~]# kubectl version
Client Version: version.Info{Major:"1", Minor:"23", GitVersion:"v1.23.1", GitCommit:"86ec240af8cbd1b60bcc4c03c20da9b98005b92e", GitTreeState:"clean", BuildDate:"2021-12-16T11:41:01Z", GoVersion:"go1.17.5", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"23", GitVersion:"v1.23.1", GitCommit:"86ec240af8cbd1b60bcc4c03c20da9b98005b92e", GitTreeState:"clean", BuildDate:"2021-12-16T11:34:54Z", GoVersion:"go1.17.5", Compiler:"gc", Platform:"linux/amd64"}

**Comment by SUMMERLm** [comment, marked as answer]
Done
should first add cluster  and then enable addon fluxcd .

**Comment by wangyikewxgm** [comment]
Enabling addon will install  CRD+controller in all cluster (hub clustet and managed Cluster). But if a cluster added in after an addon has been enabled, you should  use command  `vela addon upgrade ` to install them in new-added cluster. Please refer: https://kubevela.io/zh/docs/how-to/cli/addon/addon
